crates/ruma-state-res/src/lib.rs

```diff
@@ -167,24 +166,35 @@ fn separate<'a, Id>(
     state_sets_iter: impl Iterator<Item = &'a StateMap<Id>> + Clone,
 ) -> (StateMap<Id>, StateMap<Vec<Id>>)
 where
-    Id: Clone + Eq + 'a,
+    Id: Clone + Eq + Hash + 'a,
 {
+    let mut len: usize = 0;
+    let mut occurrences = HashMap::<_, HashMap<_, usize>>::new();
+
+    for (k, v) in state_sets_iter.clone().flat_map(|x| {
+        len += 1;
+        x
+    }) {
```

Comment:
```suggestion
    let state_sets_iter = state_sets_iter.inspect(|_| len += 1);
    for (k, v) in state_sets_iter.flatten() {
```

I find this a bit easier to follow, but don't feel strongly about it.

Reply:
I went back and forth on this in my head but I think you're right.